internal/encryption/keyprovider.go

```diff
@@ -0,0 +1,192 @@
+	if err != nil {
+		if errors.Is(err, &registry.KeyProviderNotFoundError{}) {
+			return hcl.Diagnostics{&hcl.Diagnostic{
+				Severity: hcl.DiagError,
+				Summary:  "Unknown key_provider type",
+				Detail:   fmt.Sprintf("Can not find %q", cfg.Type),
+			}}
+		}
+		return hcl.Diagnostics{&hcl.Diagnostic{
+			Severity: hcl.DiagError,
+			Summary:  fmt.Sprintf("Error fetching key_provider %q", cfg.Type),
+			Detail:   err.Error(),
+		}}
+	}
+
+	// Now that we know we have the correct Descriptor, we can decode the configuration
+	// and build the KeyProvider
+	keyProviderConfig := keyProviderDescriptor.ConfigStruct()
+
+	// Locate all the dependencies
+	deps, diags := varhcl.VariablesInBody(cfg.Body, keyProviderConfig)
+	if diags.HasErrors() {
+		return diags
+	}
+
+	// Required Dependencies
+	for _, dep := range deps {
+		// Key Provider references should be in the form key_provider.type.name
+		if len(dep) != 3 {
+			diags = append(diags, &hcl.Diagnostic{
+				Severity: hcl.DiagError,
+				Summary:  "Invalid key_provider reference",
+				Detail:   "Expected reference in form key_provider.type.name",
+				Subject:  dep.SourceRange().Ptr(),
+			})
+			continue
+		}
+
+		// TODO this should be more defensive
+		depRoot := (dep[0].(hcl.TraverseRoot)).Name
+		depType := (dep[1].(hcl.TraverseAttr)).Name
+		depName := (dep[2].(hcl.TraverseAttr)).Name
+
+		if depRoot != "key_provider" {
+			diags = append(diags, &hcl.Diagnostic{
+				Severity: hcl.DiagError,
+				Summary:  "Invalid key_provider reference",
+				Detail:   "Expected reference in form key_provider.type.name",
+				Subject:  dep.SourceRange().Ptr(),
+			})
+			continue
+		}
+
+		for _, kpc := range e.cfg.KeyProviderConfigs {
+			// Find the key provider in the config
+			if kpc.Type == depType && kpc.Name == depName {
+				depDiags := e.setupKeyProvider(kpc, stack)
+				diags = append(diags, depDiags...)
+				break
+			}
+		}
+	}
+	if diags.HasErrors() {
+		// We should not continue now if we have any diagnostics that are errors
+		// as we may end up in an inconsistent state.
+		// The reason we collate the diags here and then show them instead of showing them as they arise
+		// is to ensure that the end user does not have to play whack-a-mole with the errors one at a time.
+		return diags
+	}
+
+	// Initialize the Key Provider
+	decodeDiags := gohcl.DecodeBody(cfg.Body, e.ctx, keyProviderConfig)
+	diags = append(diags, decodeDiags...)
+	if diags.HasErrors() {
+		return diags
+	}
+
+	// Build the Key Provider from the configuration
+	keyProvider, err := keyProviderConfig.Build()
+	if err != nil {
+		return append(diags, &hcl.Diagnostic{
+			Severity: hcl.DiagError,
+			Summary:  "Unable to build encryption key data",
+			Detail:   fmt.Sprintf("%s failed with error: %s", metakey, err.Error()),
+		})
+	}
+
+	meta := e.keyProviderMetadata[metakey]
+
+	data, newmeta, err := keyProvider.Provide(meta)
+	if err != nil {
+		e.keyValues[cfg.Type][cfg.Name] = cty.UnknownVal(cty.DynamicPseudoType)
```

Comment:
@cam72cam can you add a comment here explaining why we add the dynamic type when things are unknown please?